inference-engine/samples/hetero_classification_sample_async/main.cpp

```diff
@@ -0,0 +1,324 @@
+// Copyright (C) 2018-2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+* @brief The entry point the Inference Engine sample application
+* @file classification_sample_async/main.cpp
+* @example classification_sample_async/main.cpp
+*/
+
+#include <fstream>
+#include <vector>
+#include <memory>
+#include <string>
+#include <map>
+#include <condition_variable>
+#include <mutex>
+
+#include <inference_engine.hpp>
+
+#include <format_reader_ptr.h>
+
+#include <samples/common.hpp>
+#include <samples/slog.hpp>
+#include <samples/args_helper.hpp>
+#include <samples/classification_results.h>
+
+#include <sys/stat.h>
+
+#include "classification_sample_async.h"
+
+#include <hetero/hetero_plugin_config.hpp>
+#include <ngraph/ngraph.hpp>
+#include <ngraph/op/util/op_types.hpp>
+#include <ngraph/ops.hpp>
+#include <ngraph/variant.hpp>
+#include <ngraph/opsets/opset1.hpp>
+
+using namespace InferenceEngine;
+
+void insert_noop_reshape_after(std::shared_ptr<ngraph::Node>& node, const std::string& opName) {
+    // Get all consumers for node
+    const auto consumers = node->output(0).get_target_inputs();
+
+    // Create constant with values {0, 1, 2, 3}
+    auto constant = std::make_shared<ngraph::op::Constant>(
+                ngraph::element::i64, ngraph::Shape{node->get_shape().size()}, std::vector<size_t>{0, 1, 2, 3});
+    constant->set_friendly_name(opName + "_const");
+    constant->get_rt_info()["affinity"] = std::make_shared<ngraph::VariantWrapper<std::string>>("VPUX");
+
+    // Create noop transpose node
+    auto transpose = std::make_shared<ngraph::opset1::Transpose>(node, constant);
+    transpose->set_friendly_name(opName);
+    transpose->get_rt_info()["affinity"] = std::make_shared<ngraph::VariantWrapper<std::string>>("VPUX");
+
+    // Reconnect all consumers to the new node
+    for (auto input : consumers) {
+        input.replace_source_output(transpose);
+    }
+}
+
+// Sample function that cuts network in half
+void assignAffinities(InferenceEngine::CNNNetwork& network, const std::string& firstDevice,
+                      const std::string& secondDevice, std::string splitLayer) {
+    slog::info << "Manual distribution logic is used\n";
+
+    auto ngraphFunction = network.getFunction();
+
+    auto orderedOps = ngraphFunction->get_ordered_ops();
+    auto lastSubgraphLayer =
+            std::find_if(begin(orderedOps), end(orderedOps), [&](const std::shared_ptr<ngraph::Node>& node) {
+                return splitLayer == node->get_friendly_name();
+            });
+
+    if (lastSubgraphLayer == end(orderedOps)) {
+        slog::err << "Splitting layer \"" << splitLayer << "\" was not found.";
+    }
+
+    // with VPUX plugin, also add temporary SW layer at the end of the subnetwork
+    if (firstDevice == "VPUX") {
+        splitLayer = "last_reshape_layer";
+        insert_noop_reshape_after(*lastSubgraphLayer, splitLayer);
```

Comment:
why do we need it? can this be handled in VPUX plugin itself?

Reply:
VPUX plugin does not "just work" with Hetero plugin. This sample is an example of how to hack it to make it work. It should be handled inside VPUX plugin but right now it is not.